src/Compilers/CSharp/Test/Semantic/Semantics/BetterCandidates.cs

```diff
@@ -0,0 +1,433 @@
+        [Fact]
+        public void TestAmbiguousContext01()
+        {
+            var source =
+@"public class Color
+{
+    public void M()
+    {
+        Color Color = null;
+        Color.M(null);
+    }
+
+    public static int M(A a) => 1;
+    public int M(B b) => 2;
+}
+public class A {}
+public class B {}
+";
+            CreateCompilationWithoutBetterCandidates(source, options: TestOptions.ReleaseDll).VerifyDiagnostics(
+                // (6,15): error CS0121: The call is ambiguous between the following methods or properties: 'Color.M(A)' and 'Color.M(B)'
+                //         Color.M(null);
+                Diagnostic(ErrorCode.ERR_AmbigCall, "M").WithArguments("Color.M(A)", "Color.M(B)").WithLocation(6, 15)
+                );
+            CreateCompilationWithBetterCandidates(source, options: TestOptions.ReleaseDll).VerifyDiagnostics(
+                // (6,15): error CS0121: The call is ambiguous between the following methods or properties: 'Color.M(A)' and 'Color.M(B)'
+                //         Color.M(null);
+                Diagnostic(ErrorCode.ERR_AmbigCall, "M").WithArguments("Color.M(A)", "Color.M(B)").WithLocation(6, 15)
+                );
+        }
+
+        //When a method group contains some generic methods whose type parameters do not satisfy their constraints, these members are removed from the candidate set.
+        [Fact]
+        public void TestConstraintFailed01()
+        {
+            var source =
+@"public class Program
+{
+    static void Main()
+    {
+        M(new A(), 0);
+    }
+
+    static void M<T>(T t1, int i) where T: B { System.Console.WriteLine(1); }
+    static void M<T>(T t1, short s) { System.Console.WriteLine(2); }
+}
+public class A {}
+public class B {}
+";
+            CreateCompilationWithoutBetterCandidates(source, options: TestOptions.ReleaseExe).VerifyDiagnostics(
+                // (5,9): error CS0311: The type 'A' cannot be used as type parameter 'T' in the generic type or method 'Program.M<T>(T, int)'. There is no implicit reference conversion from 'A' to 'B'.
+                //         M(new A(), 0);
+                Diagnostic(ErrorCode.ERR_GenericConstraintNotSatisfiedRefType, "M").WithArguments("Program.M<T>(T, int)", "B", "T", "A").WithLocation(5, 9)
+                );
+            var compilation = CreateCompilationWithBetterCandidates(source, options: TestOptions.ReleaseExe).VerifyDiagnostics(
+                );
+            CompileAndVerify(compilation, expectedOutput: "2");
+        }
+
+        //When a method group contains some generic methods whose type parameters do not satisfy their constraints, these members are removed from the candidate set.
+        // Test that this permits overload resolution to use type parameter constraints "as a tie-breaker" to guide overload resolution.
+        [Fact]
+        public void TestConstraintFailed02()
```

Comment:
I think we need couple more tests (i think they will pass, but in case of future regressions):
1) one candidate is generic, but candidate fails constraints, while another overload requires a conversion. Used to be an error, second should be picked now.
2) one candidate is generic without constraints, but we pass a ref-struct to it, which cannot be a generic type arg, another candidate requires a conversion and now works.

Reply:
Such tests have been added in iteration 5.

---
In reply to: [168863794](https://github.com/dotnet/roslyn/pull/24756#discussion_r168863794) [](ancestors = 168863794)